client/state/user-settings/actions.js

```diff
@@ -0,0 +1,116 @@
+/**
+ * External dependencies
+ */
+import debugFactory from 'debug';
+
+/**
+ * Internal dependencies
+ */
+import {
+	USER_SETTINGS_REQUEST,
+	USER_SETTINGS_SAVE,
+	USER_SETTINGS_UPDATE,
+	USER_SETTINGS_UNSAVED_CLEAR,
+	USER_SETTINGS_UNSAVED_SET,
+	USER_SETTINGS_UNSAVED_REMOVE,
+} from 'state/action-types';
+
+const debug = debugFactory( 'calypso:user:settings' );
+
+/**
+ * Fetch user settings from WordPress.com API and store them in UserSettings instance
+ * @returns {Function} Action thunk
+ */
+export function fetchUserSettings() {
+	return {
+		type: USER_SETTINGS_REQUEST
+	};
+}
+
+/**
+ * Post settings to WordPress.com API at /me/settings endpoint
+ *
+ * @param {Object} settingsOverride - default settings object
+ * @return {Function} Action thunk
+ */
+export function saveUserSettings( settingsOverride ) {
+	return {
+		type: USER_SETTINGS_SAVE,
+		settingsOverride
+	};
+}
+
+/**
+ * Returns an action object signalling the settings have been received from server.
+ *
+ * @param  {Object} settingValues Setting values (the subset of keys to be updated)
+ * @return {Object}               Action object
+ */
+export function updateUserSettings( settingValues ) {
+	return {
+		type: USER_SETTINGS_UPDATE,
+		settingValues
+	};
+}
+
+export function cancelPendingEmailChange() {
+	return {
+		type: USER_SETTINGS_SAVE,
+		settingsOverride: { user_email_change_pending: false }
+	};
+}
+
+/**
+ * Handles the storage and removal of changed setting that are pending
+ * being saved to the WPCOM API.
+ *
+ * @param  {String} settingName - setting name
+ * @param  {*} value - setting value
+ * @return {Function} Action thunk that returns updating successful response
+ */
+export function updateUserSetting( settingName, value ) {
+	return ( dispatch, getState ) => {
+		const { settings } = getState().userSettings;
+
+		if ( ! settings || settings[ settingName ] === undefined ) {
+			debug( settingName + ' does not exist in user-settings data module.' );
+			return false;
+		}
+
+		/*
+		 * If the two match, we don't consider the setting "changed".
+		 * user_login is a special case since the logic for validating and saving a username
+		 * is more complicated.
+		 */
+		if ( settings[ settingName ] === value && 'user_login' !== settingName ) {
+			debug( 'Removing ' + settingName + ' from changed settings.' );
+			dispatch( removeUnsavedUserSetting( settingName ) );
+		} else {
+			dispatch( updateUnsavedUserSetting( settingName, value ) );
+		}
+
+		return true;
+	};
```

Comment:
This action thunk has nothing to do with the data layer. It's an action that sets the unsaved value of a setting. The logic depends on the current state that's being read with the `getState` function available to the thunk.

This is the price one needs to pay for having independent reducers that are composed with `combineReducers`. If the next state of state node `unsavedSettings` depends on state node `settings`, the action needs to be implemented as a thunk. The `unsavedSettings` reducer doesn't see the state of `settings`.

The `updateUserSetting` action name is confusing - I'll rename it to `setUserSetting`.